Enable range partition with custom data distribution.

gallenvara opened the pull request:
Enable `range partition` with custom data distribution.

gallenvara committed:
Enable range partition with custom data distribution.

supermegaciaccount committed:
[FLINK-3309] [py] Resolve Maven warnings

mxm committed:
[FLINK-3383] move snapshot deployment from Travis CI to ASF Jenkins

Deployment will be handled by ASF's Jenkins instances which trigger
the deployment of snapshot versions every night. This way, we have not
only separated deployment from CI testing, but also improved the
reliability of the test and deployment infrastructure.

- remove Travis dependency from Maven deploy script

- set force-dependency version to Flink version

This assures deployment works when only credentials for the snapshot
deployment are available. Previously, we deployed snapshots with release
credentials. These credentials would be over-privileged.

- remove old credentials from Travis config file

This closes #1619.

greghogan committed:
[FLINK-3340] [runtime] Fix object juggling in drivers

A second attempt at object swapping in reduce drivers.

This closes #1626

supermegaciaccount committed:
[FLINK-3321] TupleSerializer.getLength() can return fixed-length size

This closes #1654.

supermegaciaccount committed:
[FLINK-3169] Remove Key class

This closes #1667.

greghogan committed:
[FLINK-3437] [web-dashboard] Fix UI router state for job plan

Transitive dependency EvEmitter updated to 1.0.2

This closes #1661

supermegaciaccount committed:
[FLINK-3432] Fixed ZookeeperOffsethandlerTest on Windows

This closes #1664.

supermegaciaccount committed:
[FLINK-3434] Prevent NPE in ClientFrontend#getClient()

This closes #1695

supermegaciaccount committed:
[FLINK-3449] createInput() no longer swallows exception

This closes #1675

Xazax-hun committed:
[FLINK-3422][streaming][api-breaking] Scramble HashPartitioner hashes.

mbalassi committed:
[FLINK-3422][streaming] Update tests reliant on hashing

Closes #1685

StephanEwen committed:
[hotfix] Make DataStream property methods properly Scalaesk

This also includes some minor cleanups

This closes #1689

Commit:
[FLINK-3461] [runtime] Remove redundant check in ZooKeeperLeaderElectionService

This closes #1700.

supermegaciaccount committed:
[FLINK-3491] [tests] Prevent URIException in HDFSCopyTest and disable HDFSCopyUtilitiesTest on Windows

This closes #1703

mxm committed:
[FLINK-3485][core] implement toString for SerializedListAccumulator

The change improves the display of SerializedListAccumulators in the web
interface by displaying the number of elements contained in the
accumulator instead of the hash code of the object.

This closes #1717.

supermegaciaccount committed:
[FLINK-3495] [tests] Disable RocksDB tests on Windows

This closes #1711

supermegaciaccount committed:
[FLINK-3496] Fix ML test discovery on Windows

This closes #1710.

rmetzger committed:
[FLINK-3509] Update Hadoop minor versions to latest & sync travis and release scripts

This closes #1713

vasia committed:
[hotfix] fix PageRank example to run with no args; some doc fixes

This closes #1714

uce committed:
[FLINK-3517] [dist] Only count active PIDs in start script

This closes #1716.

ChengXiangLi commented on flink-core/src/main/java/org/apache/flink/api/common/distributions/CustomDistribution.java:
```diff
@@ -0,0 +1,139 @@
+/*
```
CustomDistribution is only used in test, should we move this to test package?

ChengXiangLi commented on flink-java/src/main/java/org/apache/flink/api/java/DataSet.java:
```diff
@@ -1266,6 +1267,34 @@ public long count() throws Exception {
 	}
 
 	/**
+	 * Range-partitions a DataSet on the specified key fields with customed data distribution.
```
there seems too many partionBy related function in DataSet, i suggest you add these interfaces to DataSetUtils, which is a helper class to DataSet. Besides, you should add interface for both java and scala.

ChengXiangLi commented on flink-optimizer/src/main/java/org/apache/flink/optimizer/operators/AbstractJoinDescriptor.java:
```diff
@@ -142,9 +142,9 @@ public boolean areCompatible(RequestedGlobalProperties requested1, RequestedGlob
-				// Return false anyway now, we need both the partition key and data distribution
-				// information to make sure whether the range partitions are equivalent.
-				return false;
+				return checkEquivalentFieldPositionsInKeyFields(
```
Should this be done in the next issue you are working on? And i don't think "==" should be used here, as the data distribution from 2 input should not be the same object instance(they should be generated by deserialization most of time).

ChengXiangLi commented on flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/CustomDistributionITCase.java:
```diff
@@ -0,0 +1,111 @@
+	public void testRangeWithDistribution1() throws Exception{
+		ExecutionEnvironment env = ExecutionEnvironment.createLocalEnvironment();
+
+		DataSet<Tuple1<Integer>> input1 = env.fromElements(new Tuple1<Integer>(1),
```
The input data is too small, and the grouping key is unique, i don't think this can verify what you expect here.

stefanobaghino committed:
[FLINK-3518] Stale docs for quickstart setup

* Explicitly state to stop the local environment

This closes #1718

tillrohrmann committed:
[hotfix] [build] Disable exclusion rules when using build-jar maven profile.

This closes #1719

StephanEwen committed:
[FLINK-3520] [streaming] Periodic watermark operator emits current watermark in close()

This makes sure that for bounded data sets with watermarks, the final elements get properly reflected in window results.

uce committed:
[FLINK-3390] [runtime, tests] Restore savepoint path on ExecutionGraph restart

Temporary work around to restore initial state on failure during recovery as
required by a user. Will be superseded by FLINK-3397 with better handling of
checkpoint and savepoint restoring.

A failure during recovery resulted in restarting a job without its savepoint
state. This temporary work around makes sure that if the savepoint coordinator
ever restored a savepoint and there was no checkpoint after the savepoint,
the savepoint state will be restored again.

This closes #1720.

rmetzger committed:
[hotfix][deploy script] Deploy 'release-*' branches as well

tillrohrmann committed:
[FLINK-3511] [gelly] Introduce flink-gelly-examples module

The new flink-gelly-examples module contains all Java and Scala Gelly examples. The module
contains compile scope dependencies on flink-java, flink-scala and flink-clients so that
the examples can be conveniently run from within the IDE.

rmetzger committed:
Bump version to 1.1-SNAPSHOT

StephanEwen committed:
[FLINK-3522] [storm compat] PrintSampleStream prints a proper message when involked without arguments

aljoscha committed:
[FLINK-3521] Make Iterable part of method signature for WindowFunction

This closes #1723

StephanEwen committed:
[hotfix] Remove remaining unused files from the old standalone web client

tillrohrmann committed:
[FLINK-3511] [avro] Move avro examples to test scope

tillrohrmann committed:
[FLINK-3511] [hadoop-compatibility] Move hadoop-compatibility examples to test scope

tillrohrmann committed:
[FLINK-3511] [jdbc] Move jdbc examples to test scope and add flink-clients dependency

tillrohrmann committed:
[FLINK-3511] [nifi, elasticsearch] Move nifi and elasticsearch examples to test scope

tillrohrmann committed:
[FLINK-3511] [twitter] Move twitter examples to test scope

This closes #1725

StephanEwen committed:
[FLINK-3525] [runtime] Fix call to super.close() in TimestampsAndPeriodicWatermarksOperator

StephanEwen committed:
[license] Update LICENSE file for the latest version

StephanEwen committed:
[FLINK-3526] [streaming] Fix Processing Time Window Assigner and Trigger

This closes #1727

aljoscha committed:
[FLINK-3527] Add Scala DataStream.transform()

This implicitly adds KeyedStream.transform() and also explicitly
ConnectedStreams.transform().

This also removes the transform exclusions from the API completeness
tests.

StephanEwen committed:
[hotfix] Properly copy stream record in ReducingWindowBuffer and FoldingWindowBuffer

aljoscha committed:
[FLINK-3528] Add FoldingWindowBuffer for Non-Keyed Windows

This makes AllWindowedStream.fold() take constant space, just like the
keyed WindowOperator.

Also this adds a new test case in EventTimeAllWindowCheckpointingITCase
to verify that the FoldingWindowBuffer works.

This also renames the preexisting window buffers to ReducingWindowBuffer
and ListWindowBuffer to make the naming scheme consistent.

wuchong committed:
[docs] fix typos in Basic Concepts documentation

Closes #1730

uce committed:
[FLINK-3461] [runtime] Fix space indentation in ZooKeeperLeaderElectionService

tillrohrmann committed:
[FLINK-3532] [gelly] Fix artifact ID of flink-gelly-examples module

The current flink-gelly-examples artifact id wrongly used an underscore to separate
examples from flink-gelly. This commit replaces the underscore with an hyphen.

This closes #1731.

uce committed:
[FLINK-3535] [runtime-web] Decrease log verbosity of StackTraceSampleCoordinator

This closes #1732.

aljoscha committed:
[FLINK-3536] Make clearer distinction between event time and processing time

This brings it more in line with *ProcessingTimeWindows and makes it
clear what type of window assigner it is.

The old name, i.e. SlidingTimeWindows and TumblingTimeWindows is still
available but deprecated.

uce committed:
[FLINK-3534] [runtime] Prevent canceling Execution from failing

This closes #1735.

rmetzger committed:
[FLINK-3540] Shade org.htrace in flink-shaded-hadoop to get rid of its Guava dependency

This closes #1737

StephanEwen committed:
[docs] Update readme with current feature list and streaming example

aljoscha committed:
[FLINK-3538] [api-breaking] Streamline Scala DataStream.join/coGroup

This enforces that the user always has to specify keys for both inputs
before .window() can be called.

StephanEwen committed:
[FLINK-2788] [apis] Add TypeHint class to allow type-safe generic type parsing

This closes #1744

aljoscha committed:
[FLINK-3548] [api-breaking] Remove unnecessary generic parameter from SingleOutputStreamOperator

StephanEwen committed:
[FLINK-3550] [examples] Rework WindowJoin to properly demonstrate continuous window joins (Java + Scala)

vasia committed:
[docs] fix readme typos; use the same scala style in the examples

This closes #1743

StephanEwen committed:
[FLINK-3552] [examples] Add a properly windowed word count reading from a socket (Java + Scala)

uce committed:
[FLINK-3556] [runtime] Remove false check in HA blob store configuration

This closes #1749.

tillrohrmann committed:
[FLINK-3557] [stream, scala] Introduce secondary parameter list for fold function

The fold API call takes an initial value as well as a fold function. In Scala it is possible
to provide an anonymous function. In order to easily support multi line anonymous functions
as well as being consistent with Scala's collection API, this PR adds another parameter list
to the fold API call, which contains the fold function parameter.

Insert spaces between first parameter list and curly braces of anonymous function

This closes #1748.

StephanEwen committed:
[FLINK-3554] [streaming] Emit a MAX Watermark after finite sources finished

This closes #1750

uce committed:
[FLINK-3559] [dist] Don't print INFO if no active process

This closes #1751.

StephanEwen committed:
[hotfix] Removed deprecated and commented code in SourceFunctionTest

tillrohrmann committed:
[hotfix] [docs] Remove Java example code from Scala docs

mxm committed:
[maven] add module to force execution of Shade plugin

This ensures that all properties of the root pom are properly
resolved by running the Shade plugin. Thus, our root pom does not have
to depend on a Scala version just because it holds the Scala version
properties.

This closes #1755

wuchong committed:
[docs] Fix javascript exception caused by disqus and fix typos in cluster execution.

This closes #1756

rmetzger committed:
[hotfix] Make 'force-shading' deployable

StephanEwen committed:
[hotfix] Fix JavaDocs for SourceFunction

mxm committed:
[FLINK-3569] [build] Adjust test cases due to shading plugin base directory changes

tillrohrmann committed:
[FLINK-3570] [runtime] Use InetAddress.getLocalHost() as heuristic to find local address

The ConnectionUtils.findAddressUsingStrategy method tries to find out the local address which is
accessible by other machines of the cluster. It tries to connect to a specified address to do so.
In case that the no connection could be established, it uses an heuristic. Before it randomly
picked a NetworkInterface which is bound to an Inet4Address, not a loop back address and not a
link local address. In most cases it makes more sense to default to the
InetAddress.getLocalHost() address instead. This PR replaces the old heuristic with simply
returning the InetAddress.getLocalHost(). This of course requires that the system on which Flink
is running, is properly configured.

This closes #1758.

twalthr committed:
[FLINK-3566] [FLINK-3563] [core] TypeExtraction input type validation fixes

This closes #1759

aljoscha committed:
[FLINK-3575] [docs] Update 'Working With State' section in docs

This closes #1760

aljoscha committed:
[FLINK-3484] Add setSlotSharingGroup documentation

StephanEwen committed:
[docs] Add documentation about Event Time, Timestamp Assigners, Watermark Generation

wuchong committed:
[docs] Active tab caption is now displaying blue

This closes #1763

StephanEwen committed:
[FLINK-3578] [streaming scala] Support RichFunctions for Scala WindowFunction and AllWindowFunction

This also consolidates the various wrapper classes for fold() and reduce() functions.

This closes #1765

vasia committed:
[FLINK-3533] [docs] fix broken links to examples and cluster execution in Gelly guide

mxm committed:
[docs] update "build from source" page

uce committed:
[docs] Move connectors page to batch guide

mxm committed:
[FLINK-3585][tools] let deploy scripts work with spaces in paths

aljoscha committed:
[FLINK-3591] Replace Quickstart K-Means Example by Streaming Example

zentol committed:
[hotfix] Improve exception message in TaskManagerTest#testRemotePartitionNotFound

This closes #1658

mxm committed:
[FLINK-3883] fail immediately on deploy errors

uce committed:
[FLINK-3592] [docs] Update setup quickstart

uce committed:
[docs] Update stack figure on docs start page

uce committed:
[docs] Add first steps to docs start page

gallenvara committed:
Modify the first test of range partition with two key fields.

gallenvara committed:
Modify the test to validate the partitioning correctness.

gallenvara committed:
Modify the test with built-in data.

gallenvara committed:
Modify the test to avoid iterating.

Modify the test to avoid iterating.

gallenvara committed:
Move TestDataDist class into the test and modify the range boundary in the second test.

gallenvara committed:
Modify data distribution of the second test and check the partitions with
lower and upper boundary.
